lib/semian/error_rate_circuit_breaker.rb

```diff
@@ -0,0 +1,166 @@
+module Semian
+  class ErrorRateCircuitBreaker #:nodoc:
```

Comment:
I don't think this will work as well you think it will.

The error condition we are trying to dampen with a circuit breaker is a service timeout. Timeouts inherently take a long time so each timeout will be consuming a large portion of the allocated `window_size` to measure the percentage in.

So what does that mean? It means a statement like "Open the circuit if 90% of the requests are failed", doesn't actually work intuitively. In fact, if configured incorrectly, the circuit breaker will rarely ever open when you want it too.

For example, consider a `window_size` of 100 seconds, timeout of 10s, and average service response time is 50ms. If you're going for 90% of requests failed, it wont trigger after the intuitively feeling 9 timeouts. Instead, it will fail after 10 (which is 100%) because the 50ms service requests will fill any gap between timeouts with dozens of successful requests.

This rule extends for most percentages, even with a value of 1%, we're still not getting a behaviour we expect.

Based on https://github.com/Shopify/semian/issues/245, the goal of this PR is to conceptually simplify the configuration. I don't think that has been achieved. This still requires a deep understanding of how the circuit break works to configure correctly.

Reply:
An alternative could be to measure the response time of the service requests and calculate the percentage based on time spent making successful and failed requests rather than request count.

Reply:
Hi @damianthe. I agree, and it's something I've been mulling over to determine how to solve this. I'm coming from having used Hystrix, and they solve this by having a separate thread that monitors how long calls are taking. That approach is difficult in a multiprocess Ruby environment. I like the idea of tracking time spent in calls. While there may not be a good way to interrupt a call, we could at least track how many are effectively timing out. Let me give that some thought and see what I can come up with.

Reply:
You bring up a good example. I was picturing a multithreaded environment, and in that scenario 9 timeouts across 4 threads using a 100 second window would not be significant enough to open the circuit.

I like your idea of tracking time spent making successful vs failed requests. I'll work on an implementation that calculates that and will open when that crosses the threshold. I'll leave the ability to open based on % of failed calls as well, I think that is still useful in the scenario where the service is completely down. If that happens we don't spend much time making failed calls, but we also don't want to have 100s of failed calls in the 100 second window before opening the circuit.

I think having both gives flexibility to cover both scenarios.

Thank you for the feedback!

Reply:
Awesome!

> If that happens we don't spend much time making failed calls, but we also don't want to have 100s of failed calls in the 100 second window before opening the circuit.

I'm not sure I fully understand. If the circuit is completely down, it will be continuously in the half open state and all the calls made will be failed calls. Where does the % of failed calls circuit open  solution come into play here?

Reply:
> Awesome!
> 
> > If that happens we don't spend much time making failed calls, but we also don't want to have 100s of failed calls in the 100 second window before opening the circuit.
> 
> I'm not sure I fully understand. If the circuit is completely down, it will be continuously in the half open state and all the calls made will be failed calls. Where does the % of failed calls circuit open solution come into play here?

Consider your example of a circuit breaker with a time window  of 100 seconds, where the first 50 seconds (50%) of calls execute successfully. Now say the resource (redis) goes down completely (process dead), you'll now get near instantaneous connection refused errors. Those errors don't take up much time, and therefore contribute very little to the time spent in error. Therefore the circuit may have 100s or 1000s of errors before opening. In this example, you would have 40 seconds of near instant errors. 

I put a failing unit test in showing this example https://github.com/Shopify/semian/pull/264/files#diff-01dd7217b9f4b90e7bfd8d6c1e28686bR116

This is less of an issue with shorter time windows.

Reply:
I agree. I removed the test for now and will leave it as is.